[FSDP][Reland] Include duplicate parameters and modules when calling named_parameters and named_modules

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99448

The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

The previous PR is reverted due to some modules overwriting the signature of `named_parameters()`. This new PR adds a workaround for the case.

Differential Revision: [D45065973](https://our.internmc.facebook.com/intern/diff/D45065973/)

fegin committed:
[FSDP][Reland] Include duplicate parameters and modules when calling named_parameters and named_modules 

The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

The previous PR is reverted due to some modules overwriting the signature of `named_parameters()`. This new PR adds a workaround for the case.

Differential Revision: [D45065973](https://our.internmc.facebook.com/intern/diff/D45065973/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][Reland] Include duplicate parameters and modules when calling named_parameters and named_modules "


The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

The previous PR is reverted due to some modules overwriting the signature of `named_parameters()`. This new PR adds a workaround for the case.

Differential Revision: [D45065973](https://our.internmc.facebook.com/intern/diff/D45065973/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][Reland] Include duplicate parameters and modules when calling named_parameters and named_modules "


The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

The previous PR is reverted due to some modules overwriting the signature of `named_parameters()`. This new PR adds a workaround for the case.

Differential Revision: [D45065973](https://our.internmc.facebook.com/intern/diff/D45065973/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][Reland] Include duplicate parameters and modules when calling named_parameters and named_modules "


The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

The previous PR is reverted due to some modules overwriting the signature of `named_parameters()`. This new PR adds a workaround for the case.

Differential Revision: [D45065973](https://our.internmc.facebook.com/intern/diff/D45065973/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1111,7 +1112,9 @@ def _get_param_id_to_param_from_optim_input(
     def module_fn(module, prefix, tree_level, flat_param_to_fqn):
-        for param_name, param in module.named_parameters(recurse=False):
+        for param_name, param in _named_parameters_with_duplicates(
+            module, remove_duplicate=False
```
nit: "remove_duplicate=False" ==> "recurse=False"

awgu replied:
@fegin Is it not possible for there to be a duplicate with `recurse=False`? Is it okay that we are now using `recurse=True` with `remove_duplicate=False`?

I am wondering when we should use `_named_parameters_with_duplicates` with `remove_duplicate=False` and `recurse=True` (like here) versus `_named_parameters_with_duplicates` with just `recurse=False` (like `_init_flat_param_metadata` in flat_param.py).

fegin replied:
Good point. I added a check to avoid users using `remove_duplicate` with `_named_parameters_with_duplicates`. For `_get_flat_param_to_fqn`, `recurse` does not seem to affect the correctness but only redundant processes. So there are no UT errors. Added the `recurse=True`.

awgu commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -184,6 +186,22 @@ def _is_fsdp_flattened(tensor: torch.Tensor) -> bool:
+    kwargs["remove_duplicate"] = False
+    try:
+        ret = list(module.named_parameters(**kwargs))
+    except Exception as e:
```
Should we only catch the expected exception for when `remove_duplicate` is not supported as an argument instead of catching general `Exception`? It might not make a difference in practice due to how this `except` branch is written, but it might be a good practice.

fegin replied:
Yes, it does not matter as the exception will be raised again if the error is not what we expect. But since there is a concern about catch all issue, I'll change it to a more specific exception catching.

fegin committed:
Update on "[FSDP][Reland] Include duplicate parameters and modules when calling named_parameters and named_modules "


The default option of `named_parameters` and `named_modules` is to remove the duplicated parameters and modules. However, in FSDP, we need to know what parameters are shared. As a result, setting `remove_duplicate` to False is required in FSDP. Without setting `remove_duplicate` to False, FSDP won't be able to discover shared weights in some cases (e.g., the shared weights are in the same module or there are shared modules).

The previous PR is reverted due to some modules overwriting the signature of `named_parameters()`. This new PR adds a workaround for the case.

Differential Revision: [D45065973](https://our.internmc.facebook.com/intern/diff/D45065973/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
